components/search/search.tsx

```diff
@@ -0,0 +1,392 @@
+};
+
+const Search: React.FC<Props> = (props: Props): JSX.Element => {
+    const {actions, searchTerms} = props;
+
+    // generate intial component state and setters
+    const [focussed, setFocussed] = useState<boolean>(false);
+    const [keepInputFocussed, setKeepInputFocussed] = useState<boolean>(false);
+    const [indexChangedViaKeyPress, setIndexChangedViaKeyPress] = useState<boolean>(false);
+    const [highlightedSearchHintIndex, setHighlightedSearchHintIndex] = useState<number>(-1);
+    const [visibleSearchHintOptions, setVisibleSearchHintOptions] = useState<SearchHintOption[]>(
+        determineVisibleSearchHintOptions(searchTerms),
+    );
+
+    const suggestionProviders = [
+        new SearchDateProvider(),
+        new SearchChannelProvider(actions.autocompleteChannelsForSearch),
+        new SearchUserProvider(actions.autocompleteUsersInTeam),
+    ];
+
+    useEffect((): void => {
+        setVisibleSearchHintOptions(determineVisibleSearchHintOptions(searchTerms));
+    }, [searchTerms]);
+
+    // handle cloding of rhs-flyout
+    const handleClose = (): void => actions.closeRightHandSide();
+
+    // focus the search input
+    const handleFocus = (): void => setFocussed(true);
+
+    // release focus from the search input or unset `keepInputFocussed` value
+    // `keepInputFocussed` is used to keep the search input focussed when a
+    // user selects a suggestion from `SearchHint` with a click
+    const handleBlur = (): void => {
+        // add time out so that the pinned and member buttons are clickable
+        // when focus is released from the search box.
+        setTimeout((): void => {
+            if (keepInputFocussed) {
+                setKeepInputFocussed(false);
+            } else {
+                setFocussed(false);
+            }
+        }, 0);
+
+        updateHighlightedSearchHint(0);
+    };
+
+    const handleSearchHintSelection = (): void => {
+        setKeepInputFocussed(true);
+    };
+
+    const handleAddSearchTerm = (term: string): void => {
+        const pretextArray = searchTerms?.split(' ') || [];
+        pretextArray.pop();
+        pretextArray.push(term.toLowerCase());
+        actions.updateSearchTerms(pretextArray.join(' '));
+        updateHighlightedSearchHint(0);
+    };
+
+    const handleUpdateSearchTerms = (term: string): void => {
+        actions.updateSearchTerms(term);
+        setFocussed(true);
+    };
+
+    const handleChange = (e: ChangeEvent<HTMLInputElement>): void => {
+        const term = e.target.value;
+        actions.updateSearchTerms(term);
+    };
+
+    const updateHighlightedSearchHint = (indexDelta: number, changedViaKeyPress = false): void => {
+        let newIndex = highlightedSearchHintIndex + indexDelta;
+
+        switch (indexDelta) {
+        case 1:
+            // KEY.DOWN
+            // is it at the end of the list?
+            newIndex = newIndex === visibleSearchHintOptions.length ? 0 : newIndex;
```

Comment:
I don't see where this could help here?
It is a modulo function, that will return `NaN` when `visibleSearchHintOptions.length === 0`.
Or did you mean to use it another way?

Reply:
I meant that it would handle keeping the index between 0 and the length like this is doing but in a more concise way. I didn't realize that there's a case where the length would be 0 though.

Reply:
fair enough ...